### PR TITLE
Added xdg_surface 

### PR DESCRIPTION
**Pull request by TheNeikos**

### CODE REVIEW

**Comment by neithernut** [comment]
Mh... the setup for code generation from the XML is still missing. I'll show you how to do it on monday.

File: src/compositor/wayland/xdg_surface.c [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    uint32_t version,
+    uint32_t id,
+    struct swc_surface* surface
+) {
+    //!< @todo implement
+}
+
+
+
+
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_surface.c]: Four blank lines are three too much!

Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_surface.c]: Also check the blank lines between the functions.

File: src/compositor/wayland/xdg_shell.c [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+ws_object_type_id WS_OBJECT_TYPE_ID_WAYLAND_XDG_SHELL = {
+    .supertype  = &WS_OBJECT_TYPE_ID_OBJECT,
+    .typestr    = "ws_wayland_xdg_shell",
+
+    .hash_callback = NULL,
+    .deinit_callback = NULL,
+    .dump_callback = NULL,
+    .run_callback = NULL,
+    .cmp_callback = NULL,
+};
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_shell.c:120]: Here are two things missing: The attribute table and the function table.

File: src/compositor/wayland/xdg_shell.c [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+int
+ws_wayland_xdg_shell_init(void) {
+    bool is_static = false;
+    if (is_static) {
+        return 0;
+    }
+
+    struct wl_display* display = ws_wayland_acquire_display();
+
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_shell.c:137]: Please remove this blank line.

File: src/compositor/wayland/xdg_shell.c [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    is_static = true;
+    return 0;
+}
+
+struct ws_wayland_xdg_shell*
+ws_wayland_xdg_shell_new(
+    struct wl_resource* resource
+) {
+    struct ws_wayland_xdg_shell* self = calloc(1, sizeof(*self));
+
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_shell.c]: Please remove this blank line.

File: src/compositor/wayland/xdg_shell.c [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    if (!self) {
+        return NULL;
+    }
+
+    int retval = ws_wayland_obj_init(&self->wl_obj, resource);
+
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_shell.c]: Please remove this blank line.

File: src/compositor/wayland/xdg_shell.c [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    if (retval < 0) {
+        return NULL;
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_shell.c:165]: You lose the `struct ws_wayland_xdg_shell` here.

File: src/compositor/wayland/xdg_shell.c [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    return (struct ws_waysome_xdg_shell*) wl_resource_get_user_data(resource);
+}
+
+static void
+use_unstable_version(
+    struct wl_client* client,
+    struct wl_resource* resource,
+    int32_t version
+) {
+
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_shell.c]: Missing `//!< @todo` here.

File: src/compositor/wayland/xdg_shell.c [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    uint32_t id,
+    struct wl_resource* surface_resource,
+    struct wl_resource* parent_resource,
+    struct wl_resource* seat_resource,
+    uint32_t serial,
+    int32_t x,
+    int32_t y,
+    uint32_t flags
+) {
+
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_shell.c]: Missing `//!< @todo` here.

File: src/compositor/wayland/xdg_shell.c [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+static void pong(
+    struct wl_client* client,
+    struct wl_resource* resource,
+    uint32_t serial
+) {
+
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_shell.c]: Missing `//!< @todo` here.

File: src/compositor/wayland/xdg_shell.c [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+static void bind_shell(
+    struct wl_client* client,
+    void* data,
+    uint32_t version,
+    uint32_t id
+) {
+    version = 1; // We only implement version 1 right now
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_shell.c:247]: Missing `//!< @todo` here?

File: src/compositor/wayland/xdg_surface.c [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+    .show_window_menu       = &show_window_menu,
+    .move                   = &move,
+    .resize                 = &resize,
+    .ack_configure          = &ack_configure,
+    .set_window_geometry    = &set_window_geometry,
+    .set_maximized          = &set_maximized,
+    .unset_maximized        = &unset_maximized,
+    .set_fullscreen         = &set_fullscreen,
+    .unset_fullscreen       = &unset_fullscreen,
+    .set_minimized          = &set_minimized,
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_surface.c]: These are all function ptrs, right? Why do you have to `&` here?

File: src/compositor/wayland/xdg_surface.c [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+) {
+    struct wl_resource* resource;
+    resource = ws_wayland_client_create_resource(client, &xdg_surface_interface,
+                                                 version, id);
+    if (!resource) {
+        return NULL;
+    }
+
+    struct ws_wayland_xdg_surface* self = calloc(1, sizeof(*self));
+
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_surface.c]: Please remove this blank line.

File: src/compositor/wayland/xdg_surface.c [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+    if (!self) {
+        goto cleanup_wayland;
+        return NULL;
+    }
+
+    int retval = ws_wayland_obj_init(&self->wl_obj, resource);
+
```
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_surface.c]: Please remove this blank line.

File: src/protocol/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,19 @@
+
+
```
Comment (matthiasbeyer) [inline src/protocol/CMakeLists.txt]: Please remove these blank lines.

**Comment by matthiasbeyer** [comment]
The indentation of the XML is pretty fucked up, I guess you copy-pasted it, right? Can you please reindent it?

File: src/compositor/wayland/xdg_shell.c (hunk 2 above)
Comment (neithernut) [inline src/compositor/wayland/xdg_shell.c:120]: Will not work without rebasing. It's based on a commit in which the function table wasn't yet integrated.

**Comment by neithernut** [comment]
I would not reindent it, but keep it the way it is, as it _may_ be updated by C&Ping over a new version.

**Comment by TheNeikos** [comment]
Yeah, I took the xml as is.

**Comment by matthiasbeyer** [comment]
@neithernut made a good point.

File: src/compositor/wayland/xdg_shell.c (hunk 2 above)
Comment (matthiasbeyer) [inline src/compositor/wayland/xdg_shell.c:120]: So he has to rebase...

File: src/compositor/wayland/xdg_shell.c (hunk 10 above)
Comment (TheNeikos) [inline src/compositor/wayland/xdg_shell.c:247]: Well, there is nothing left to do, but we might support other versions later down the line

**Comment by TheNeikos** [comment]
Okay, I fixed all your comments, as you have not made them on the now modified lines they still appear as unresolved. I trust you to change that.

**Comment by matthiasbeyer** [comment]
Feel free to do some squashing! I guess we will pair-review this evening!

**Comment by neithernut** [comment]
ACK.

**Comment by matthiasbeyer** [comment]
ACK.

**Pull request merged**